### PR TITLE
Stricter types

### PR DESCRIPTION
**Pull request by victorlin**
## Description of proposed changes

While working on #1073, I thought it was reasonable to avoid casting and non-null assertions across the entire codebase. This is an attempt to enforce it and address outstanding violations.

## Related

Follow-up to https://github.com/nextstrain/nextstrain.org/pull/1073#discussion_r1843065163

## Checklist

- [x] Merge base PR #1073
- [x] Checks pass
- [x] Check if changes affect the [resource index JSON revision](https://docs.nextstrain.org/projects/nextstrain-dot-org/en/latest/resource-collection.html#resource-index-revisions)
- [ ] Post-merge: do the same in Auspice

### CODE REVIEW

File: static-site/src/components/Groups/Tiles/index.tsx [hunk 1]
```diff
@@ -7,6 +7,7 @@ import { UserContext } from "../../../layouts/userDataWrapper";
 import { GroupTile } from "./types";
 import { Group } from "../types";
 import { ExpandableTiles } from "../../ExpandableTiles";
+import { InternalError } from "../../ListResources/errors";
```
Comment (victorlin) [inline static-site/src/components/Groups/Tiles/index.tsx:10]: Hmm, I think this warrants a separate error boundary for `static-site/src/components/Groups/Tiles/index.tsx`.

File: static-site/src/components/Groups/Tiles/index.tsx [hunk 2]
```diff
@@ -25,8 +26,13 @@ function createGroupTiles(groups: Group[], colors = [...theme.titleColors]): Gro
...
     .map((group) => {
-      const groupColor = colors[0]!;
-      colors.push(colors.shift()!);
+      if (colors[0] === undefined) {
+        throw new InternalError("Colors are missing.");
+      }
+      const groupColor = colors[0];
+
+      // Rotate the colors
+      colors.push(colors.shift()!);  // eslint-disable-line @typescript-eslint/no-non-null-assertion
```
Comment (genehack) [inline static-site/src/components/Groups/Tiles/index.tsx:35]: can you alleviate the need for the non-null assert by repeating the `if (colors[0] === undefined) {throw}` check prior to this?

File: static-site/src/components/ListResources/IndividualResource.tsx [hunk 3]
```diff
@@ -139,9 +139,17 @@ export const IndividualResource = ({
       throw new InternalError("ref must be defined and the parent must be a div (IndividualResourceContainer).");
      }
 
+     // The type of ref.current.parentNode is ParentNode which does not have an
+     // offsetTop property I don't think there is a way to appease the
```
Comment (genehack) [inline static-site/src/components/ListResources/IndividualResource.tsx:143]:
```suggestion
     // offsetTop property. I don't think there is a way to appease the
```

**Review by genehack — APPROVED** [review]

File: static-site/src/components/ListResources/IndividualResource.tsx (hunk 3 above)
Comment (victorlin) [reply static-site/src/components/ListResources/IndividualResource.tsx]: Sharp eyes! Done: https://github.com/nextstrain/nextstrain.org/blob/e62d01c2b918108851e23ee1d5849501a4bec8e9/static-site/src/components/ListResources/IndividualResource.tsx#L143

File: static-site/src/components/Groups/Tiles/index.tsx (hunk 1 above)
Comment (victorlin) [reply static-site/src/components/Groups/Tiles/index.tsx]: Added 0af3c3a8 + 96f3018d to address this. Not with a separate error boundary but by reusing the existing one.

**Comment by victorlin** [comment]
@genehack could you re-review? I've pushed 4982fa8833ee524ed2c510b9b5768e51c2c94246 to account for the OpenGraph addition in 639eae0d567cdc77d8b675ce54d50c4847254c30.

**Review by genehack — APPROVED** [review]

**Pull request merged**